bluesky/protocols.py

```diff
@@ -227,14 +243,62 @@ def subscribe(self, function: Callable) -> Union[NoReturn, None]:
         """
         ...
 
-    def clear_sub(self, function: Callable) -> Union[NoReturn, None]:
+    def clear_sub(self, function: Callable) -> None:
         """Remove a subscription."""
         ...
 
 
+class Subscription(Protocol):
+    def __enter__(self) -> "Subscription":
+        """Returns self"""
+        return self
+
+    def __exit__(self, type, value, traceback):
+        """Calls self.close()"""
+        self.close()
+
+    def close(self):
+        """Closes connection"""
+        ...
+
+
+@runtime_checkable
+class Monitorable(Protocol):
+    """Async compatible version of Subscribable with tweaked interface"""
+    # TODO: monitor() returns a Subscription, do we like the naming?
+    def monitor(
+        self, callback: Callable[[Dict[str, EventData]], Union[None, Awaitable]]
+    ) -> Subscription:
+        """Monitor for updates in value of a device.
+
+        Callback should be called in the asyncio event loop thread with
+        EventData dict like that of read() on each update
+
+        Expected usage:
+
+            with signal.monitor(queue.put):
+                event_datas = await queue.get()
+                do_something_with(event_datas)
+                # subscriptions marked for closing here
+
+        Or:
+
+            sub = signal.monitor(callback)
+            # Sometime later
+            sub.close()
+
+        Needed for :doc:`monitored <async>`.
+        """
+        ...
+
+
 @runtime_checkable
 class Checkable(Protocol):
-    def check_value(self, value: Any) -> Union[NoReturn, None]:
+    # TODO: should this be allowed to do I/O to do this check? If so,
+    # then this should only be used after stage() as that might change
+    # some params. If not, then it will need some cached limits to check
+    # against which still probably come from stage.
```

Comment:
This is a tricky problem that bit us with pymalcolm, we change settings as part of setting up a device for a scan, but want to check if a scan config is valid before that happens. I thought we might be able to use saved settings to validate against, but if we have a `stage` plan then maybe we could run that and parse the `set` messages to use those values?

Reply:
Probably best to make this `SyncOrAsync[None]` but put in the docs that no signals should be written to as this can be called during a scan